sql/core/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -1234,6 +1234,8 @@ class Column(val expr: Expression) extends Logging {
    */
   def over(): Column = over(Window.spec)
 
+  def copy(field: String, value: Column): Column = withExpr(StructCopy(expr, field, value.expr))
```

Comment:
Some things to consider about the API:

- How is custom metadata associated with the updated field?
- How can a field be deleted?
- How can a field be added?
    - When a field is added, where does it go in the schema? The only logical place is at the end but that may not be what's desired in some cases.

Simply for discussion purposes (overloaded methods are not shown):

```scala
class Column(val expr: Expression) extends Logging {

  // ...

  // matches Dataset.schema semantics; errors on non-struct columns
  def schema: StructType

  // matches Dataset.select() semantics, errors on non-struct columns
  // '* support allows multiple new fields to be added easily, saving cumbersome repeated withColumn() calls
  def select(cols: Column*): Column

  // matches Dataset.withColumn() semantics of add or replace
  def withColumn(colName: String, col: Column): Column

  // matches Dataset.drop() semantics
  def drop(colName: String): Column

}
```

The benefit of the above API is that it unifies manipulating top-level & nested columns, which I would argue is very desirable. The addition of `schema` and `select()` allows for nested field reordering, casting, etc., which is important in data exchange scenarios where field position matters.

/cc @rxin

Reply:
BTW, to fully unify top-level and nested column manipulation, we should add `Dataset.cast(schema: StructType)` as `cast()` can be used effectively to change the schema of a struct, e.g.,

```scala
spark.range(1, 2).toDF("x")
  .withColumn("st", struct('x.as("x"), 'x.cast(DoubleType).as("y")))
  .select('st.cast(new StructType().add("y", LongType).add("x", IntegerType)))
```

For `Dataset`, outside of `DataFrameReader.schema()`, which may be too early in the transformation flow, there is no easy way for end users to change a schema. Experienced Spark developers will have no problem coming up with `spark.createDataFrame(df.rdd, newSchema)` but that's neither consistent with how `cast` works nor chainable as a transformation unless used with `transform()` which takes us quite far away from simple Spark user code.

Reply:
> Can .cast() be expressed via the .copy() method if the former one will support add/delete/update operations?

Are you suggesting `Dataset` adds `copy()` as opposed to `cast()`?

Otherwise, certainly, `Dataset.cast()` can be implemented with `Dataset.select()` but it's not easy even in the simple cases and becomes noticeably more complicated when you consider correctly setting nullability & metadata in the edge cases